### PR TITLE
Http_Parser to Request Conversion, NotFoundHandler, Minor RequestHandler Refactor

### PR DESCRIPTION
**Pull request by kevinazhu**
1. Removed Http_Parser and replaced with Request class which follows in-class specification.
2. Add 404 NotFoundHandler
3. Simplified RequestHandler base by removing unnecessary parameter

### CODE REVIEW

File: request.cc [hunk 1]
```diff
@@ -1,57 +1,60 @@
...
 }
 
-std::string HttpParser::get_body() const
+std::string Request::body() const
 {
     return body_;
 }
 
-bool HttpParser::parse_first_line(std::string line)
+bool Request::parse_first_line(std::string line)
```
Comment (jteo1) [inline request.cc]: Make line parameter a const?

File: file_handler.h [hunk 2]
```diff
@@ -2,16 +2,18 @@
...
+#include <unordered_map>
 #include "request_handler.h"
 #include "config_parser.h"
-#include "http_parser.h"
+#include "request.h"
 
 class FileHandler : public RequestHandler {
 public:
 	FileHandler(std::string directory);
 	// data is the full http request
```
Comment (jteo1) [inline file_handler.h]: Make directory parameter const?

File: echo_handler.cc [hunk 3]
```diff
@@ -4,11 +4,11 @@
 #include "request_handler.h"
 #include "echo_handler.h"
 
-std::string EchoHandler::GenerateResponse(const HttpParser& headerInfo, const std::string& requestData) const
+std::string EchoHandler::GenerateResponse(std::unique_ptr<Request> &request) const
```
Comment (dstephan1) [inline echo_handler.cc]: This doesn't have to be a unique_ptr. Changing this also breaks the API we're supposed to match. 

I recommend const Request&. The call can be changed like this: response = handler->GenerateResponse(*requestptr);

File: not_found_handler.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+#include "request_handler.h"
+#include "config_parser.h"
+#include "request.h"
+
+class NotFoundHandler : public RequestHandler {
+public:
+	
+	virtual std::string GenerateResponse(std::unique_ptr<Request> &request) const;
+
```
Comment (dstephan1) [inline not_found_handler.h]: TODO:
change std::string to Response class,
rename GenerateResponse to HandleRequest
constructor magic

Comment (kevinazhu) [inline not_found_handler.h]: Will add Response object and constructor magic in another branch/pull request

Comment (dstephan1) [inline not_found_handler.h]: sounds good

**Pull request merged**